File: task7/find_key.go
package main

import (
	"bufio"
	"crypto/aes"
	"crypto/cipher"
	"encoding/hex"
	"fmt"
	"os"
)

// Pad the key to 16 bytes with '#'
func padKey(key string) []byte {
	paddedKey := make([]byte, 16)
	copy(paddedKey, key)
	for i := len(key); i < 16; i++ {
		paddedKey[i] = '#'
	}
	return paddedKey
}

// Perform AES-128-CBC decryption
func aesDecrypt(ciphertext []byte, key []byte, iv []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}

	if len(ciphertext) < aes.BlockSize {
		return nil, fmt.Errorf("ciphertext too short")
	}

	mode := cipher.NewCBCDecrypter(block, iv)
	plaintext := make([]byte, len(ciphertext))
	mode.CryptBlocks(plaintext, ciphertext)

	// Remove padding
	paddingLen := int(plaintext[len(plaintext)-1])
	if paddingLen > aes.BlockSize || paddingLen > len(plaintext) {
		return nil, fmt.Errorf("padding size error")
	}
	plaintext = plaintext[:len(plaintext)-paddingLen]

	return plaintext, nil
}

func main() {
	ciphertextHex := "44b7e8d03b633600ea3aaabf42807627fabce9af94e4ba8c46da8efb360d0d2457ccb004914cb8dc99d59584d50eae4f"
	ivHex := "010203040506070809000a0b0c0d0e0f"
	plaintext := "my stories dont end until I stop running"

	ciphertext, _ := hex.DecodeString(ciphertextHex)
	iv, _ := hex.DecodeString(ivHex)

	// Open the wordlist file
	wordlist, err := os.Open("words.txt")
	if err != nil {
		fmt.Println("Error opening wordlist:", err)
		return
	}
	defer wordlist.Close()

	var word string
	key := make([]byte, 16)

	scanner := bufio.NewScanner(wordlist)
	for scanner.Scan() {
		word = scanner.Text()
		if len(word) == 0 {
			continue
		}

		// Pad the word to 16 characters using '#'
		key = padKey(word)

		decryptedtext, err := aesDecrypt(ciphertext, key, iv)
		if err != nil {
			continue
		}

		if string(decryptedtext) == plaintext {
			fmt.Println("Key found:", word)
			break
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading wordlist:", err)
	}
}

File: task7/find_key.c
#include <stdio.h>
#include <string.h>
#include <openssl/evp.h>
#include <openssl/aes.h>

// Following code is made with help of https://docs.openssl.org/1.1.1/man3/EVP_EncryptInit/#examples and CursorAI

#define KEY_SIZE 16 // AES-128 requires 16 bytes key
#define IV_SIZE 16  // AES-128-CBC uses a 16-byte IV (Initail Vector)

// Helper function to pad the key with '#' to 16 characters
void pad_key(char *key, unsigned char *padded_key)
{
    int key_len = strlen(key);
    memset(padded_key, '#', KEY_SIZE);
    memcpy(padded_key, key, key_len);
}

// Function to perform AES-128-CBC decryption
int aes_decrypt(unsigned char *ciphertext, int ciphertext_len, unsigned char *key, unsigned char *iv, unsigned char *plaintext)
{
    EVP_CIPHER_CTX *ctx;
    int len;
    int plaintext_len;

    // Create and initialize the context
    if (!(ctx = EVP_CIPHER_CTX_new()))
        return -1;

    // Initialize decryption operation with AES-128-CBC
    if (1 != EVP_DecryptInit_ex(ctx, EVP_aes_128_cbc(), NULL, key, iv))
        return -1;

    // Decrypt the ciphertext
    if (1 != EVP_DecryptUpdate(ctx, plaintext, &len, ciphertext, ciphertext_len))
        return -1;
    plaintext_len = len;

    // Finalize the decryption
    if (1 != EVP_DecryptFinal_ex(ctx, plaintext + len, &len))
        return -1;
    plaintext_len += len;

    // Clean up
    EVP_CIPHER_CTX_free(ctx);

    return plaintext_len;
}

int main()
{
    // Given data
    unsigned char ciphertext[] = {0x76, 0x4a, 0xa2, 0x6b, 0x55, 0xa4, 0xda, 0x65, 0x4d, 0xf6, 0xb1, 0x9e, 0x4b, 0xce, 0x00, 0xf4,
                                  0xed, 0x05, 0xe0, 0x93, 0x46, 0xfb, 0x0e, 0x76, 0x25, 0x83, 0xcb, 0x7d, 0xa2, 0xac, 0x93, 0xa2};
    unsigned char iv[] = {0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff, 0x00, 0x99, 0x88, 0x77, 0x66, 0x55, 0x44, 0x33, 0x22, 0x11};
    char plaintext[] = "This is a top secret.";

    unsigned char decryptedtext[128];
    int decryptedtext_len;

    // Open the wordlist file
    FILE *wordlist = fopen("words.txt", "r");
    if (wordlist == NULL)
    {
        perror("Error opening wordlist");
        return 1;
    }

    char word[KEY_SIZE];
    unsigned char key[KEY_SIZE];

    // Loop through each word in the wordlist
    while (fgets(word, sizeof(word), wordlist))
    {
        // Remove newline character from word
        word[strcspn(word, "\n")] = 0;

        // Pad the word to 16 characters using '#'
        pad_key(word, key);

        // Decrypt the ciphertext
        decryptedtext_len = aes_decrypt(ciphertext, sizeof(ciphertext), key, iv, decryptedtext);
        if (decryptedtext_len < 0)
        {
            continue; // Decryption failed, try next key
        }

        // Null-terminate the decrypted string
        decryptedtext[decryptedtext_len] = '\0';

        // Compare decrypted text with the plaintext
        if (strcmp((char *)decryptedtext, plaintext) == 0)
        {
            printf("Key found: %s\n", word);
            break;
        }
    }

    fclose(wordlist);
    return 0;
}
